glib-macros/src/gboxed_derive.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use proc_macro2::TokenStream;
+use proc_macro_error::abort_call_site;
+use quote::quote;
+
+use crate::utils::parse_type_name;
+
+pub fn impl_gboxed(input: &syn::DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    let gtype_name = match parse_type_name(&input, "gboxed") {
+        Ok(v) => v,
+        Err(e) => abort_call_site!(
+            "{}: derive(GBoxed) requires #[gboxed(type_name = \"BoxedTypeName\")]",
+            e
+        ),
+    };
+
+    quote! {
+        impl BoxedType for #name {
+            const NAME: &'static str = #gtype_name;
+
+            fn get_type() -> glib::Type {
+                static mut TYPE_: glib::Type = glib::Type::Invalid;
+                static ONCE: ::std::sync::Once = ::std::sync::Once::new();
+
+                ONCE.call_once(|| {
+                    let type_ = glib::subclass::register_boxed_type::<Self>();
+                    unsafe {
+                        TYPE_ = type_;
+                    }
+                });
+
+                unsafe { TYPE_ }
+            }
+        }
+
+        impl glib::StaticType for #name {
+            fn static_type() -> glib::Type {
+                <#name as glib::subclass::boxed::BoxedType>::get_type()
+            }
+        }
+
+        impl glib::value::SetValue for #name {
+            unsafe fn set_value(value: &mut glib::value::Value, this: &Self) {
+                let ptr: *mut #name = Box::into_raw(Box::new(this.clone()));
+                glib::gobject_sys::g_value_take_boxed(
+                    glib::translate::ToGlibPtrMut::to_glib_none_mut(value).0,
+                    ptr as *mut _,
+                );
+            }
+        }
+
+        impl glib::value::SetValueOptional for #name {
+            unsafe fn set_value_optional(value: &mut glib::value::Value, this: Option<&Self>) {
+                let this = this.expect("None not allowed");
```

Comment:
This is interesting now. With configuration on the derive macro we could allow nullable and non-nullable types :)

Reply:
Something like that (see new top commit)?